Jenkinsfile

```diff
@@ -13,46 +12,110 @@ pipeline {
         }
       }
     }
-    stage('Build') {
-      steps {
-        ansiColor('xterm') {
-          sh '''
-            make deps  -B
-            make build -B -j4
-          '''
+    stage('Makefile-based') {
+      agent {
+        dockerfile {
+          additionalBuildArgs '--build-arg USER_ID=$(id -u) --build-arg GROUP_ID=$(id -g)'
+          reuseNode true
         }
       }
-    }
-    stage('OCaml Backend') {
-      steps {
-        ansiColor('xterm') {
-          sh '''
-            nprocs=$(nproc)
-            [ "$nprocs" -gt '4' ] && nprocs=4
-            make TEST_CONCRETE_BACKEND=ocaml test-execution -j"$nprocs"
-          '''
+      stages {
+        stage('Build') {
+          steps {
+            ansiColor('xterm') {
+              checkout scm
+              sh '''
+                make deps -B
+                make build -B -j4
+              '''
+            }
+          }
+        }
+        stage('Exec OCaml') {
+          steps {
+            ansiColor('xterm') {
+              sh '''
+                nprocs=$(nproc)
+                [ "$nprocs" -gt '4' ] && nprocs=4
+                make TEST_CONCRETE_BACKEND=ocaml test-execution -j"$nprocs"
+              '''
+            }
+          }
+        }
+        stage('Exec Java') {
+          steps {
+            ansiColor('xterm') {
+              sh '''
+                nprocs=$(nproc)
+                [ "$nprocs" -gt '4' ] && nprocs=4
+                make TEST_CONCRETE_BACKEND=java test-execution -j"$nprocs"
+              '''
+            }
+          }
+        }
+        stage('Proofs Java') {
+          steps {
+            ansiColor('xterm') {
+              sh '''
+                nprocs=$(nproc)
+                [ "$nprocs" -gt '4' ] && nprocs=4
+                make test-proof -j"$nprocs"
+              '''
+            }
+          }
         }
       }
     }
-    stage('Java Backend') {
-      steps {
-        ansiColor('xterm') {
-          sh '''
-            nprocs=$(nproc)
-            [ "$nprocs" -gt '4' ] && nprocs=4
-            make TEST_CONCRETE_BACKEND=java test-execution -j"$nprocs"
-          '''
+    stage('KNinja-based') {
+      agent {
+        dockerfile {
+          additionalBuildArgs '--build-arg USER_ID=$(id -u) --build-arg GROUP_ID=$(id -g)'
+          reuseNode true
         }
       }
-    }
-    stage('Test Proofs') {
-      steps {
-        ansiColor('xterm') {
-          sh '''
-            nprocs=$(nproc)
-            [ "$nprocs" -gt '4' ] && nprocs=4
-            make test-proof -j"$nprocs"
-          '''
+      stages {
+        stage('Build') {
+          steps {
+            ansiColor('xterm') {
+              checkout scm
+              sh '''
+                ./build --opamroot ~/.opam wasm-java wasm-ocaml wasm-haskell
```

Comment:
why does the user need to specify this --opamroot flag?

Reply:
Default is `.build/opam` (inside the repo), which is nice for user builds so that we get local opam roots.

But for CI server, we want `$HOME/.opam` instead, so we can take advantage of dockerfile cacheing.

Reply:
I'm not really sure why you would ever want a per-repository opam root for the ocaml backend.

Reply:
For people who only install one semantics and otherwise don't care about K and have other OCaml projects which they don't want messed with. It localizes all changes to the repo, which is a clear win.

Reply:
It is not required. I prefer that builds of a repo don't affect the users state outside of that repo, so KNinja defaults to setting `$OPAMROOT` to `.build/opam`. This means that users don't need to worry about K projects interfering with e.g. `coq` installs that need another `ocamlc` version. For the CI, we set it to `~/.opam` so that opam dependencies can be installed by the `Dockefile` saving build time.

Reply:
Oh I see what you're saying. The real solution to that problem is to invoke eval `opam config env --switch 4.06.1+k` in the shell that executes ocaml commands, and then simply don't set our custom switch as the default when we create it.

Reply:
Building an entire opam switch takes significant time and I have no intention of building it once for every project we build. You can do something cheap for now if you want, but it needs to be addressed before this kninja stuff could possibly be ready to be used as anything other than an experiment.

Reply:
We are no longer using OPAMROOT, but setting `OPAMSWITCH = 4.06.1+k` as default switch in the upstream KNinja repo.